### DISCUSSION TITLE
Диакритические символы

### DISCUSSION THREAD

**Discussion by unxed** [elfmz/far2l#1281] [Q&A]
Хорошо бы нам начать поддерживать диакритические (и вообще комбинированные) символы на уровне экранного буфера. То есть, уйти от концепции «1 ячейка экрана = 1 кодовая точка unicode». На самом деле, кодовых точек юникода на ячейку может быть сколько угодно. А вот как бы такое устроить без серьезного проседания производительности?

Работает же это во всяких терминалках юникодных как-то — mc в GNOME Terminal нормально показывает файл `ÀÀÀÀÀÀÀÀÀÀÀÀÀÀÀÀ.txt`, а вот far2l так не может.

**Comment by elfmz** [comment]
Надо рефакторить код фара чтоб он воспринимал терминал не как сетку wchar-ов, а учитывал длину строки в каждом конкретном месте. А хаки типа https://github.com/elfmz/far2l/pull/1279 добавлять не хочется.
А чтобы сильно рефакторить код фара надо сделать мало мальские функциональные автотесты, юниттесты на готовый код писать бессмысленно, особенно когда частенько не понимаешь как оно работает.
Я себе это представляю как отдельную апликуху предоставляющую псевдотерминал для запуска far2l --tty (ну хотябы для начала один бэкенд тестировать), и которая шлет в него нажатия кнопок и анализирует что фар там на экране рисует.
Причем, чтобы упростить себе жизнь, можно сразу делать этакий макро-рекордер, на работающий вне фара. Чтоб можно было натыкать вручную сценарий, который потом бы натыкивался автоматически при билде.
Короче, тянет на отдельный прожект))

**Reply by unxed** [reply]
А как вы понимаете это «учитывал длину строки» на практике? Что будет представлять собой экранный буфер при такой архитектуре?

Окей, если автоматизированное тестирование — главная загвоздка, подумаю в эту сторону. Для начала можно просто спросить на stackoverflow, и посмотреть, чего люди посоветуют. Не факт, что будет толк, но за спрос, как говорится, денег не берут :)

https://stackoverflow.com/questions/71237385/is-there-a-way-to-automatically-test-interactive-command-line-console-linux-ap

https://stackoverflow.com/questions/71237212/what-is-the-maximum-number-of-unicode-combined-characters-that-may-be-needed-to

**Reply by unxed** [reply]
А может прямо на базе фара такое сделать? У него есть встроенный терминал уже, есть запись макросов. Нужно только добавить скармливание макросов дочернему фару с некоторой задержкой; возможность проверять, чего там в выдаче; ну и автоматизацию запуска/отчётов.

**Comment by unxed** [comment]
@elfmz для автоматизированного тестирования интерактивных консольных приложений есть тулза `expect`, она делает прямо в точности то, что вы хотите! Только что проверил, работает с far2l чудесно.

`sudo apt install expect`
```
#!/bin/bash
OUTPUT=$(expect -c '
#!/usr/bin/expect
# setting log file name
log_file far2l_test.log
# setting timeout
set timeout 2
# spawning far2l (expect will not work without --mortal switch)
spawn far2l --tty --mortal
# "Press <ENTER> if tired of watching this message"? ok, simulate Enter
send -- "\x0d"
# Do nothing for 1 second
sleep 1
# Loaded?
expect "Help"
# Simulating Shift+F4
send -- "\x1bO2S"
# Awaiting "Editor" string
expect "Editor"
# Simulating F10
send -- "\x1b\[21~"
# Simulating F10 again to exit far2l
send -- "\x1b\[21~"
# Awaiting "Quit" string
expect {
	timeout {puts "Timeout"; exit 1; }
	eof     {puts "Far2l crashed"; exit 2; }
	"Quit"
}
# Simulating Enter
send -- "\x0d"
# Awaiting far2l to terminate
expect eof
')
echo "Expect's return value: $?"
```

Можно подумать о неком конвертере макросов фара в expect-скрипты для упрощения подготовки тестов, это должно быть несложно. Возможно, понадобится чутка доработать сам макро-рекордер, чтоб он в каменты макросов сохранял характерные строки для expect'а (например, можно после каждого нажатия клавиши сохранять в камент макроса первый найденный в экранном буфере текстовый фрагмент, отсутствовавший на экране до нажатия клавиши).

**Reply by unxed** [reply]
Можно, собственно, на базе самого фара сделать такой конвертер из макроса в expect скрипт, используя функции KeyNameToKey и VT_TranslateSpecialKey.

**Comment by unxed** [comment]
Сделал небольшой патчик к far2l, позволяющий при записи макроса одновременно создавать скрипт для `expect`. Пока поддерживаются не все клавиши, это чисто Proof-of-Concept.

При этом уже получается записывать вполне работающие скрипты, пример ниже.

```
diff --git a/far2l/src/macro/macro.cpp b/far2l/src/macro/macro.cpp
index 8703e8da..bf6a901b 100644
--- a/far2l/src/macro/macro.cpp
+++ b/far2l/src/macro/macro.cpp
@@ -5174,6 +5174,82 @@ void KeyMacro::SaveMacros(BOOL AllSaved)
 #endif
 		cfg_writer.SetString("Sequence", MacroLIB[I].Src);
 
+		// --
+		wchar_t * temp = (wchar_t*)malloc((wcslen(MacroLIB[I].Src)+1)*sizeof(wchar_t));
+		wcsncpy ( temp, MacroLIB[I].Src, wcslen(MacroLIB[I].Src) + 1 );
+
+		char *file;
+		char *fileName = "/macro.sh";
+		file = (char*)malloc(strlen(getenv("HOME")) + strlen(fileName) + 1);
+		strcpy(file, getenv("HOME"));
+		strcat(file, fileName);
+
+		FILE *fp = fopen(file, "w");
+
+		wchar_t* buffer;
+		wchar_t* token = std::wcstok(temp, L" ", &buffer);
+		while (token) {
+
+			int keycode = KeyNameToKey(token);
+
+			// debug
+			// fprintf (fp, "%ls - %d\n", token, keycode);
+
+			if (keycode == 0x0d) // or enter will be interpreted as ctrl-enter by far2l
+				fprintf (fp, "sleep 0.01\n");
+
+			fprintf (fp, "send -- \"");
+
+			if (keycode == 8) { // VK_BACK
+				fprintf (fp, "\\x7f");
+			} else if (keycode == 27) { // ESC
+				fprintf (fp, "\\x1b\\x1b");
+			} else if (keycode == 0x0d) { // enter
+				fprintf (fp, "\\x0d");
+			} else if (keycode == 34) { // double quote
+				fprintf (fp, "\\\"");
+			} else if (wcslen(token) == 1) { // other alphanumeric
+				fprintf (fp, "%lc", keycode);
+			} else {
+
+				int keycode_new = keycode;
+				if (keycode & KEY_CTRL) keycode_new -= KEY_CTRL;
+				if (keycode & KEY_ALT) keycode_new -= KEY_ALT;
+				if (keycode & KEY_SHIFT) keycode_new -= KEY_SHIFT;
+				if (keycode & KEY_ALTDIGIT) keycode_new -= KEY_ALTDIGIT;
+				if (keycode & KEY_RSHIFT) keycode_new -= KEY_RSHIFT;
+
+				const char* out = 
+					VT_TranslateSpecialKey(
+						keycode_new,
+						(keycode & KEY_CTRL),
+						(keycode & KEY_ALT) || (keycode & KEY_ALTDIGIT),
+						(keycode & KEY_SHIFT) || (keycode & KEY_RSHIFT),
+						0,
+						keycode_new
+					);
+
+				if (out && strcmp(out, "(null)") != 0) {
+					if (out[0] == 27) { // esc seq
+						if (out[1] == 91) { // left sqare bracket
+							out++; fprintf (fp, "\\x1b\\[%s", ++out);
+						} else {
+							fprintf (fp, "\\x1b%s", ++out);
+						}
+					} else {
+						fprintf (fp, "%s", out);
+					}
+				}
+			}
+			fprintf (fp, "\"\n");
+
+			token = std::wcstok(nullptr, L" ", &buffer);
+		}
+
+		fclose(fp);
+		free(temp);
+		// --
+
 		if (MacroLIB[I].Description)
 			cfg_writer.SetString("Description", MacroLIB[I].Description);
 		else
diff --git a/far2l/src/macro/macro.hpp b/far2l/src/macro/macro.hpp
index 4c3d5b22..b24d8c2c 100644
--- a/far2l/src/macro/macro.hpp
+++ b/far2l/src/macro/macro.hpp
@@ -33,6 +33,8 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 
+extern const char *VT_TranslateSpecialKey(const WORD key, bool ctrl, bool alt, bool shift, unsigned char keypad, WCHAR uc);
+
 #include "syntax.hpp"
 #include "tvar.hpp"
 #include "macroopcode.hpp"
```

А вот пример записанного скрипта (шапка и хвост вручную добавлены):

```
#!/usr/bin/expect
# setting log file name
log_file far2l_test.log
# setting timeout
set timeout 20
# spawning far2l (expect will not work without --mortal switch)
spawn far2l --tty --mortal --nodetect

send -- "\x1bO3Q"
send -- "/"
send -- "\x1bO3P"
send -- "/"
send -- "\x1be"
send -- "\x1bt"
send -- "\x1bc"
sleep 0.01
send -- "\x0d"
send -- "\x1bh"
send -- "\x1bo"
send -- "\x1bs"
send -- "\x1bt"
send -- "\x1bs"
send -- "\x1bOS"
send -- "\x1b\[1;5H"
send -- "\x1b\[18~"
send -- "i"
send -- "p"
send -- "v"
send -- "6"
sleep 0.01
send -- "\x0d"
send -- "\x1b\[1;6C"

# Awaiting far2l to terminate
expect eof
```

Для понимания, вот эквивалентный key_macros.ini
```
[KeyMacros/Shell/1]
DisableOutput=0x1
Sequence=AltF2 / AltF1 / AltE AltT AltC Enter AltH AltO AltS AltT AltS F4 CtrlHome F7 i p v 6 Enter CtrlShiftRight
```

**Comment by unxed** [comment]
Можно сделать даже ещё изящнее: запускать фар внутри фара, и дублировать в файл все esc-последовательности ввода, генерируемые «внешним» фаром в ответ на нажатия клавиш в режиме far2l extensions. А потом при тестировании скармливать их фару через ту же expect. Так, по идее, можно вообще любые сочетания клавиш эмулировать, а не только доступные в обычных терминалках. Можно даже фиксировать интервалы времени между нажатиями, и вставлять в скрипты sleep, чтобы оно точь в точь по времени воспроизводилось.
